src/serlio/prtMaterial/ArnoldMaterialNode.cpp

```diff
@@ -92,147 +105,97 @@ MStatus ArnoldMaterialNode::compute(const MPlug& plug, MDataBlock& data) {
-	sb.declString(L"$normalMapConvertNode");
-	sb.declString(L"$colorMapBlendNode");
-	sb.declString(L"$dirtMapBlendNode");
-	sb.declString(L"$opacityMapBlendNode");
-	sb.declString(L"$specularMapBlendNode");
-	sb.declString(L"$emissiveMapBlendNode");
-	sb.declString(L"$roughnessMapBlendNode");
-	sb.declString(L"$metallicMapBlendNode");
-	sb.declString(L"$uvTrafoNode");
+	MaterialUtils::MaterialCache matCache = MaterialUtils::getMaterialsByStructure(materialStructure);
+	MaterialNameSource nameSource;
+
+	MELScriptBuilder scriptBuilder;
+	scriptBuilder.declString(L"$shadingGroup");
+	scriptBuilder.declString(L"$shaderNode");
+	scriptBuilder.declString(L"$mapFile");
+	scriptBuilder.declString(L"$mapNode");
+	scriptBuilder.declString(L"$bumpLuminanceNode");
+	scriptBuilder.declString(L"$bumpValueNode");
+	scriptBuilder.declString(L"$displacementNode");
+	scriptBuilder.declString(L"$normalMapConvertNode");
+	scriptBuilder.declString(L"$colorMapBlendNode");
+	scriptBuilder.declString(L"$dirtMapBlendNode");
+	scriptBuilder.declString(L"$opacityMapBlendNode");
+	scriptBuilder.declString(L"$specularMapBlendNode");
+	scriptBuilder.declString(L"$emissiveMapBlendNode");
+	scriptBuilder.declString(L"$roughnessMapBlendNode");
+	scriptBuilder.declString(L"$metallicMapBlendNode");
+	scriptBuilder.declString(L"$uvTrafoNode");
 
 	for (adsk::Data::Handle& inMatStreamHandle : *inMatStream) {
 		if (!inMatStreamHandle.hasData())
 			continue;
 
-		if (!inMatStreamHandle.usesStructure(*fStructure))
+		if (!inMatStreamHandle.usesStructure(*materialStructure))
 			continue;
 
-		MaterialInfo matInfo(inMatStreamHandle);
-
-		if (!inMatStreamHandle.setPositionByMemberName(gPRTMatMemberFaceStart.c_str())) {
+		if (!inMatStreamHandle.setPositionByMemberName(gPRTMatMemberFaceStart.c_str()))
 			continue;
-		}
 		const int faceStart = *inMatStreamHandle.asInt32();
 
-		if (!inMatStreamHandle.setPositionByMemberName(gPRTMatMemberFaceEnd.c_str())) {
+		if (!inMatStreamHandle.setPositionByMemberName(gPRTMatMemberFaceEnd.c_str()))
 			continue;
-		}
 		const int faceEnd = *inMatStreamHandle.asInt32();
 
-		// get unique name
-		std::wstring shaderName(L"serlioGeneratedArnoldMaterialSh");
-		std::wstring shadingGroupName(L"serlioGeneratedArnoldMaterialSg");
+		auto createShadingEngine = [this, &nameSource, &materialStructure, &scriptBuilder,
+		                            &inMatStreamHandle](const MaterialInfo& matInfo) {
+			auto newMaterialName = nameSource.getUniqueName(MATERIAL_BASE_NAME);
+			const std::wstring& shadingEngineName = newMaterialName.first;
+			const std::wstring& shaderName = newMaterialName.second;
+
+			synchronouslyCreateShadingEngine(shadingEngineName);
+			MaterialUtils::assignMaterialMetadata(materialStructure, inMatStreamHandle, shadingEngineName);
+			appendToMaterialScriptBuilder(scriptBuilder, matInfo, shaderName, shadingEngineName);
+			LOG_INF << "new arnold shading engine: " << shadingEngineName << " (surface shader: " << shaderName << ")";

```

Comment:
do we want to keep the logging (same below)?

Reply:
remind me to remove it before merging ;-)

Reply:
nice